File: pfchangs/app/src/main/java/testclient/filter/AMCL.java
package testclient.filter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;

import org.opencv.core.Point3;

import edu.wpi.first.math.filter.LinearFilter;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import testclient.Constants;
import testclient.MathX;
import testclient.wrappers.TagDistance;

// FIXME check if EVERYTHING (besides LL campose/botpose) is in radians
public class AMCL {
    private HashMap<Integer, TagDistance> distances = new HashMap<>();

    private Point3 motionDelta = new Point3(); // robot frame motion odometry <-- maybe not?

    private ArrayList<Particle> particles = new ArrayList<Particle>();
    private Particle bestEstimate = new Particle(0, 0, 0, 0);
    private Particle meanEstimate = new Particle(0, 0, 0, 0);
    private Particle weightedAverage = new Particle();

    private double mGaussX, mGaussY, mGaussW;
    private double vGaussX, vGaussY, vGaussW;

    private double mclASlow, mclAFast, mclWFast, mclWSlow; // AMCL vars

    private boolean useHeading;

    private double nParticles;

    private boolean useAdaptiveParticles;
    private double cf, dist;

    public AMCL() {
        useHeading = true;
    }

    public void init() {
        Random r = new Random();

        nParticles = Constants.FilterConstants.NUM_PARTICLES;
        for (int i = 0; i < nParticles; ++i) {
            particles.add(new Particle(
                    r.nextDouble(-Constants.FIELD_WIDTH_OFFSET,
                            Constants.FIELD_WIDTH - Constants.FIELD_WIDTH_OFFSET),
                    r.nextDouble(-Constants.FIELD_HEIGHT_OFFSET,
                            Constants.FIELD_HEIGHT - Constants.FIELD_HEIGHT_OFFSET),
                    r.nextDouble(Math.PI * 2),
                    1 / nParticles));
        }

        mGaussX = 0.1; // meters, 0.15
        mGaussW = 0.1; // radians, 0.2
        vGaussW = 0.15; // (degrees) radians, 0.15
        vGaussY = 0.1; // meters, 0.1
        mGaussY = 0.1; // meters, 0.15
        mclASlow = 0.01; // 0.01
        useAdaptiveParticles = false;
        mclAFast = 0.1; // 0.1
        vGaussX = 0.1; // meters, 0.1
    }

    /**
     * @param angle             The particle's angle
     * @param setpoint          The known angle to compare <code>angle</code>
     *                          against (usually from Limelight measurements)
     * @param angleIsRadians    Whether or not <code>angle</code> is in radians
     * @param setpointIsRadians Whether or not <code>setpoint</code> is in radians
     * @return
     */
    private double headingErr(double angle, double setpoint, boolean angleIsRadians, boolean setpointIsRadians) {
        if (!angleIsRadians) {
            angle = Math.toRadians(angle);
        }

        if (!setpointIsRadians) {
            setpoint = Math.toRadians(setpoint);
        }

        angle = Math.abs(angle);
        setpoint = Math.abs(setpoint);

        angle %= (2 * Math.PI);
        if (angle < 0)
            angle += 2 * Math.PI;
        return MathX.Gaussian(setpoint, vGaussW, angle);
    }

    /**
     * @param angle    The particle's angle in radians
     * @param setpoint The known angle to compare <code>angle</code> against in
     *                 radians
     * @return
     */
    private double headingErr(double angle, double setpoint) {
        return headingErr(angle, setpoint, true, true);
    }

    /**
     * @param x Robot-relative x translation
     * @param y Robot-relative y translation
     * @param w Rotation in radians
     */
    public void updateOdometry(double x, double y, double w) {
        motionDelta.x = x;
        motionDelta.y = y;
        motionDelta.z = w;

        updateMotion();
    }

    public void setNoise(
            double mGaussX,
            double mGaussY,
            double mGaussW,
            double vGaussX,
            double vGaussY) {
        this.mGaussX = mGaussX;
        this.mGaussY = mGaussY;
        this.mGaussW = mGaussW;
        this.vGaussX = vGaussX;
        this.vGaussY = vGaussY;
    }

    public void setMCLParams(
            double mclAFast,
            double mclASlow) {
        this.mclAFast = mclAFast;
        this.mclASlow = mclASlow;
    }

    public void resetMCL() {
        Random r = new Random();

        nParticles = Constants.FilterConstants.NUM_PARTICLES;
        for (int i = 0; i < nParticles; ++i) {
            particles.add(new Particle(
                    r.nextDouble(-Constants.FIELD_WIDTH_OFFSET,
                            Constants.FIELD_WIDTH - Constants.FIELD_WIDTH_OFFSET),
                    r.nextDouble(-Constants.FIELD_HEIGHT_OFFSET,
                            Constants.FIELD_HEIGHT - Constants.FIELD_HEIGHT_OFFSET),
                    r.nextDouble(Math.PI * 2),
                    1 / nParticles));
        }
    }

    private void updateMotion() {
        Random r = new Random();

        double dx = motionDelta.x;
        double dy = motionDelta.y;
        double dw = motionDelta.z;
        for (var p : particles) {
            p.x += dx + r.nextGaussian(0, mGaussX);
            p.y += dy + r.nextGaussian(0, mGaussY);
            p.w += dw + r.nextGaussian(0, mGaussW);

            p.w %= (2 * Math.PI);
            if (p.w < 0)
                p.w += 2 * Math.PI;

            if (p.x > Constants.FIELD_WIDTH - Constants.FIELD_WIDTH_OFFSET)
                p.x = Constants.FIELD_WIDTH - Constants.FIELD_WIDTH_OFFSET;
            else if (p.x < 0 - Constants.FIELD_WIDTH_OFFSET)
                p.x = 0 - Constants.FIELD_WIDTH_OFFSET;

            if (p.y > Constants.FIELD_HEIGHT - Constants.FIELD_HEIGHT_OFFSET)
                p.y = Constants.FIELD_HEIGHT - Constants.FIELD_HEIGHT_OFFSET;
            else if (p.y < 0 - Constants.FIELD_HEIGHT_OFFSET)
                p.y = 0 - Constants.FIELD_HEIGHT_OFFSET;
        }
    }

    public void tagScanning(int id, double[] dists, double[] campose) {
        for (int i = 0; i < dists.length; ++i) {
            if (dists[i] > 0) {
                distances.put(i + 1,
                        new TagDistance(Constants.TAG_ARR[i].x, Constants.TAG_ARR[i].y, dists[i]));
            }
        }

        updatePerceptionPoints(id, campose);
    }

    private void updatePerceptionPoints(int id, double[] campose) {
        int numPoints = distances.size();
        double sumWeight = 0;
        double wAvg = 0;

        for (var p : particles) {
            double prob = 1;
            double cmpsProb = 0;

            if (numPoints > 0) {
                for (var d : distances.values()) {
                    double tagDist = d.distance + MathX.normalDistribution(0, Math.hypot(vGaussX, vGaussY));
                    double particleDistance = Math.hypot(p.x - d.x, p.y - d.y);
                    double distanceDiff = Math.abs(particleDistance - tagDist);
                    prob *= MathX.Gaussian(0, Math.hypot(vGaussX, vGaussY), distanceDiff);
                }

                p.weight = prob;

                if (useHeading && id > 1) {
                    cmpsProb = headingErr(p.w,
                            (campose[2] + Math.toRadians(Constants.TAG_ARR[id - 1].z)) % (2 * Math.PI) +
                                    MathX.normalDistribution(0, vGaussW));
                    p.weight *= cmpsProb;
                }
                sumWeight += p.weight;
            }
        }

        for (var p : particles) {
            wAvg += p.weight / nParticles;
            p.weight /= sumWeight;
        }

        mclWSlow += mclASlow * (wAvg - mclWSlow);
        mclWFast += mclAFast * (wAvg - mclWFast);

        computeWeightedAverage();
        lowVarResampling();
    }

    private void lowVarResampling() {
        ArrayList<Particle> newParticles = new ArrayList<>();
        Random random = new Random();
        double resetProb = Math.max(0, 1 - (mclWFast / mclWSlow));
        double r = random.nextDouble() * (1 / nParticles);
        double c = particles.get(0).weight;
        bestEstimate = particles.get(0);
        int id = 0;

        double meanX = 0;
        double meanY = 0;
        double orient = 0;

        for (int j = 0; j < Constants.FilterConstants.NUM_PARTICLES; ++j) {
            SmartDashboard.putNumber("resetprob", resetProb);
            if (MathX.bernoulliDistribution(resetProb)) {
                newParticles.add(new Particle(
                        random.nextDouble(-Constants.FIELD_WIDTH_OFFSET,
                                Constants.FIELD_WIDTH - Constants.FIELD_WIDTH_OFFSET),
                        random.nextDouble(-Constants.FIELD_HEIGHT_OFFSET,
                                Constants.FIELD_HEIGHT - Constants.FIELD_HEIGHT_OFFSET),
                        random.nextDouble(Math.PI * 2),
                        1 / Constants.FilterConstants.NUM_PARTICLES));
            } else {
                double U = r + ((double) j / nParticles);
                while (U > c) {
                    id += 1;
                    c += particles.get(id).weight;
                }
                if (particles.get(id).weight > bestEstimate.weight) {
                    bestEstimate = particles.get(id).clone();
                }
                newParticles.add(particles.get(id).clone());
            }
        }

        particles = newParticles;

        dist = 0;
        cf = 0;
        bestEstimate = getBestEstimate();
        double xBest = bestEstimate.x;
        double yBest = bestEstimate.y;

        for (var p : particles) {
            meanX += p.x;
            meanY += p.y;
            orient += p.w;
            p.weight = 1 / Constants.FilterConstants.NUM_PARTICLES;

            double tmpX = xBest - p.x;
            double tmpY = yBest - p.y;
            dist += Math.hypot(tmpX, tmpY);
        }

        meanX /= Constants.FilterConstants.NUM_PARTICLES;
        meanY /= Constants.FilterConstants.NUM_PARTICLES;
        orient /= Constants.FilterConstants.NUM_PARTICLES;

        meanEstimate.x = meanX;
        meanEstimate.y = meanY;
        meanEstimate.w = orient;

        if (useAdaptiveParticles) {
            cf = (dist / nParticles) / 18.38067;
            if (cf >= 0.08)
                nParticles = cf * Constants.FilterConstants.NUM_PARTICLES;
            else
                nParticles = Constants.FilterConstants.MIN_PARTICLES;
        } else {
            nParticles = Constants.FilterConstants.NUM_PARTICLES;
        }
    }

    public Particle getBestEstimate() {
        bestEstimate = particles.get(0);

        for (int j = 0; j < nParticles; ++j) {
            if (particles.get(j).weight > bestEstimate.weight) {
                bestEstimate = particles.get(j);
            }

        }

        return bestEstimate.clone();
    }

    public Particle getAverageEstimate() {
        meanEstimate = new Particle(0, 0, 0, 0);

        double meanX = 0, meanY = 0, meanW = 0;
        for (var p : particles) {
            meanX += p.x;
            meanY += p.y;
            meanW += p.w;
        }

        meanEstimate.x = meanX / particles.size();
        meanEstimate.y = meanY / particles.size();
        meanEstimate.w = meanW / particles.size();
        meanEstimate.weight = 1 / particles.size();

        return meanEstimate.clone();
    }

    public Particle computeWeightedAverage() {
        double meanX = 0, meanY = 0, meanW = 0;
        for (var p : particles) {
            meanX += p.x * p.weight;
            meanY += p.y * p.weight;
            meanW += p.w * p.weight;
        }

        weightedAverage = new Particle(meanX, meanY, meanW, 0);
        return weightedAverage.clone();
    }

    public Particle getWeightedAverage() {
        return weightedAverage.clone();
    }

    public Particle getFilteredAverage() {
        Particle tmp = getAverageEstimate();
        LinearFilter xFilter = LinearFilter.movingAverage(10);
        LinearFilter yFilter = LinearFilter.movingAverage(10);
        LinearFilter wFilter = LinearFilter.movingAverage(10);

        return new Particle(xFilter.calculate(tmp.x), yFilter.calculate(tmp.y), wFilter.calculate(tmp.w), 0);
    }

    public Particle getFilteredWeightedAverage() {
        Particle tmp = getWeightedAverage();
        LinearFilter xFilter = LinearFilter.movingAverage(10);
        LinearFilter yFilter = LinearFilter.movingAverage(10);
        LinearFilter wFilter = LinearFilter.movingAverage(10);

        return new Particle(xFilter.calculate(tmp.x), yFilter.calculate(tmp.y), wFilter.calculate(tmp.w), 0);
    }

    public void outputNParticles() {
        SmartDashboard.putNumber("Number of particles", nParticles);
    }

    @Override
    public String toString() {
        String res = "";
        for (var p : particles)
            res += p + "\n";
        return res;
    }
}

File: re-fchangs/rust/src/amcl.rs
use std::{collections::HashMap, f64::consts::PI};

use crate::{
    constants::*, math_util::*, particle::Particle, point3::Point3, tag_distance::TagDistance,
};

pub struct Amcl {
    distances: HashMap<i32, TagDistance>,

    motion_delta: Point3,

    particles: Vec<Particle>,
    best_estimate: Particle,
    mean_estimate: Particle,
    weighted_average: Particle,

    motion_gauss_x: f64,
    motion_gauss_y: f64,
    motion_gauss_w: f64,

    vision_gauss_x: f64,
    vision_gauss_y: f64,
    vision_gauss_w: f64,

    mcl_aslow: f64,
    mcl_afast: f64,
    mcl_wslow: f64,
    mcl_wfast: f64,

    use_heading: bool,
}

impl Amcl {
    pub fn new() -> Amcl {
        Amcl {
            distances: HashMap::new(),
            motion_delta: Point3(0.0, 0.0, 0.0),
            particles: Vec::new(),
            best_estimate: Particle::default(),
            mean_estimate: Particle::default(),
            weighted_average: Particle::default(),
            motion_gauss_x: 0.1,
            motion_gauss_y: 0.1,
            motion_gauss_w: 0.15,
            vision_gauss_x: 0.1,
            vision_gauss_y: 0.1,
            vision_gauss_w: 0.15,
            mcl_aslow: 0.01,
            mcl_afast: 0.1,
            mcl_wslow: 0.0,
            mcl_wfast: 0.0,
            use_heading: true,
        }
    }

    pub fn init(&mut self) {
        for _ in 0..NUM_PARTICES {
            self.particles.push(Particle(
                uniform_random_distribution(-FIELD_WIDTH_OFFSET, FIELD_WIDTH - FIELD_WIDTH_OFFSET),
                uniform_random_distribution(
                    -FIELD_HEIGHT_OFFSET,
                    FIELD_HEIGHT - FIELD_HEIGHT_OFFSET,
                ),
                uniform_random_distribution(0.0, PI * 2.0),
                1.0 / NUM_PARTICES as f64,
            ));
        }
    }

    fn heading_err<const ANGLE_IS_RADIANS: bool, const SETPOINT_IS_RADIANS: bool>(
        mut angle: f64,
        mut setpoint: f64,
        sigma: f64,
    ) -> f64 {
        if !ANGLE_IS_RADIANS {
            angle = angle.to_radians();
        }

        if !SETPOINT_IS_RADIANS {
            setpoint = setpoint.to_radians();
        }

        angle = angle.abs();
        setpoint = setpoint.abs();

        angle %= 2.0 * PI;
        while angle < 0.0 {
            angle += 2.0 * PI;
        }
        gaussian(setpoint, sigma, angle)
    }

    pub fn update_odometry(&mut self, x: f64, y: f64, w: f64) {
        *self.motion_delta.x() = x;
        *self.motion_delta.y() = y;
        *self.motion_delta.z() = w;
        self.update_motion();
    }

    fn update_motion(&mut self) {
        let dx = *self.motion_delta.x();
        let dy = *self.motion_delta.y();
        let dw = *self.motion_delta.z();

        for p in &mut self.particles {
            *p.x() += dx + normal_distribution(0.0, self.motion_gauss_x);
            *p.y() += dy + normal_distribution(0.0, self.motion_gauss_y);
            *p.w() += dw + normal_distribution(0.0, self.motion_gauss_w);

            *p.w() %= 2.0 * PI;
            if *p.w() < 0.0 {
                *p.w() += 2.0 * PI;
            }

            if *p.x() > FIELD_WIDTH - FIELD_WIDTH_OFFSET {
                *p.x() = FIELD_WIDTH - FIELD_WIDTH_OFFSET;
            } else if *p.x() < -FIELD_WIDTH_OFFSET {
                *p.x() = -FIELD_WIDTH_OFFSET;
            }

            if *p.y() > FIELD_HEIGHT - FIELD_HEIGHT_OFFSET {
                *p.y() = FIELD_HEIGHT - FIELD_HEIGHT_OFFSET;
            } else if *p.y() < -FIELD_HEIGHT_OFFSET {
                *p.y() = -FIELD_HEIGHT_OFFSET;
            }
        }
    }

    pub fn tag_scanning(&mut self, id: i32, dists: [f64; NUM_TAGS], campose: [f64; 3]) {
        self.distances.clear();
        for i in 0..dists.len() {
            if dists[i] > 0.0 {
                self.distances.insert(
                    i as i32 + 1,
                    TagDistance(TAG_ARRAY[i].x(), TAG_ARRAY[i].y(), dists[i]),
                );
            }
        }

        self.update_perception(id, campose);
    }

    fn update_perception(&mut self, id: i32, campose: [f64; 3]) {
        let num_points = self.distances.len();
        let mut sum_weight: f64 = 0.0;
        let mut w_avg: f64 = 0.0;

        for p in &mut self.particles {
            let mut prob: f64 = 1.0;

            if num_points > 0 {
                for d in &mut self.distances.values_mut() {
                    let tag_dist: f64 = *d.distance()
                        + normal_distribution(
                            0.0,
                            f64::hypot(self.vision_gauss_x, self.vision_gauss_y),
                        );
                    let particle_distance: f64 = f64::hypot(*p.x() - *d.x(), *p.y() - *d.y());
                    let distance_diff: f64 = f64::abs(particle_distance - tag_dist);
                    prob *= gaussian(
                        0.0,
                        f64::hypot(self.vision_gauss_x, self.vision_gauss_y),
                        distance_diff,
                    );
                }

                *p.weight() = prob;

                if self.use_heading && id > 1 {
                    let cmps_prob = Self::heading_err::<true, true>(
                        *p.w(),
                        (campose[2] + TAG_ARRAY[id as usize - 1_usize].z()).to_radians()
                            % (2.0 * PI)
                            + normal_distribution(0.0, self.vision_gauss_w),
                        self.vision_gauss_w,
                    );

                    *p.weight() *= cmps_prob;
                }

                sum_weight += *p.weight();
            }
        }

        for p in &mut self.particles {
            w_avg += *p.weight() / NUM_PARTICES as f64;
            *p.weight() /= sum_weight;
        }

        self.mcl_wslow += self.mcl_aslow * (w_avg - self.mcl_wslow);
        self.mcl_wfast += self.mcl_afast * (w_avg - self.mcl_wfast);

        self.compute_weighted_average();
        self.low_var_resampling();
    }

    fn low_var_resampling(&mut self) {
        let reset_prob: f64 = f64::clamp(1.0 - (self.mcl_wfast - self.mcl_wslow), 0.0, 1.0);
        let r = uniform_random_distribution(0.0, 1.0 / NUM_PARTICES as f64);
        let mut new_particles: Vec<Particle> = Vec::new();
        let mut c = *self.particles.first_mut().unwrap().weight();
        let mut id = 0;
        self.best_estimate = *self.particles.first().unwrap();

        for j in 0..NUM_PARTICES {
            if bernoulli_distribution(reset_prob) {
                new_particles.push(Particle(
                    uniform_random_distribution(
                        -FIELD_WIDTH_OFFSET,
                        FIELD_WIDTH - FIELD_WIDTH_OFFSET,
                    ),
                    uniform_random_distribution(
                        -FIELD_HEIGHT_OFFSET,
                        FIELD_HEIGHT - FIELD_HEIGHT_OFFSET,
                    ),
                    uniform_random_distribution(0.0, PI * 2.0),
                    1.0 / NUM_PARTICES as f64,
                ));
            } else {
                // `u` is the threshold that `c` must cross in order for a particle to be selected for the next generation.
                // each iteration of the inner loop, the weight of the current particle (indexed by `id`) is added to
                // `c`. the particle that puts `c` "over the top" is selected for resampling.
                // `u` increases with each iteration of the outer loop, ensuring that different particles get selected
                // with each outer loop iteration. (i.e., if `u` stayed the same, the same particle would put us over
                // the top every iteration, despite there probably being more particles that we haven't looked at yet.)
                // this is how we make sure the higher-weighted particles are selected for the next generation.
                // of course, there is a chance that a high-weight particle puts us just under the threshold, then a
                // lower-weight particle is the one that puts us over the top. there's not much anyone can do in that case.
                // finding a better resampling algorithm is left as an exercise to the reader.
                let u = r + (j as f64 / NUM_PARTICES as f64);
                while u > c {
                    id += 1;
                    c += *self.particles.get_mut(id).unwrap().weight();
                }

                if self.particles.get_mut(id).unwrap().weight() > self.best_estimate.weight() {
                    self.best_estimate = *self.particles.get(id).unwrap();
                }

                new_particles.push(*self.particles.get(id).unwrap());
            }
        }

        self.particles = new_particles;

        // Recompute best and average estimates after resampling
        self.get_best_estimate();
        self.get_average_estimate();
    }

    pub fn get_best_estimate(&mut self) -> Particle {
        self.best_estimate = *self.particles.first().unwrap();

        for i in 0..NUM_PARTICES {
            if self.particles.get_mut(i).unwrap().weight() > self.best_estimate.weight() {
                self.best_estimate = *self.particles.get(i).unwrap();
            }
        }

        self.best_estimate
    }

    pub fn get_average_estimate(&mut self) -> Particle {
        self.mean_estimate = Particle::default();

        let (mut mean_x, mut mean_y, mut mean_w) = (0.0, 0.0, 0.0);
        for p in &mut self.particles {
            mean_x += *p.x();
            mean_y += *p.y();
            mean_w += *p.w();
        }

        *self.mean_estimate.x() = mean_x / NUM_PARTICES as f64;
        *self.mean_estimate.y() = mean_y / NUM_PARTICES as f64;
        *self.mean_estimate.w() = mean_w / NUM_PARTICES as f64;

        self.mean_estimate
    }

    pub fn compute_weighted_average(&mut self) -> Particle {
        let (mut mean_x, mut mean_y, mut mean_w) = (0.0, 0.0, 0.0);
        for p in &mut self.particles {
            mean_x += *p.x() * *p.weight();
            mean_y += *p.y() * *p.weight();
            mean_w += *p.w() * *p.weight();
        }

        self.weighted_average = Particle(mean_x, mean_y, mean_w, 0.0);
        self.weighted_average
    }

    pub fn get_weighted_average(&mut self) -> Particle {
        self.weighted_average
    }
}

impl Default for Amcl {
    fn default() -> Self {
        Self::new()
    }
}
